angr/knowledge_plugins/labels.py

```diff
@@ -1,66 +1,219 @@
+        addr = self.get_addr(name)
+        if addr is None:
+            raise KeyError(name)
+        return addr
+
+    #
+    #   ...
+    #
+
+    def set_label(self, addr, name, ns=_global_ns):
+        """Label address `addr` as `name` within namespace `namespace`
+        
+        :param addr:        The address to put a label on.
+        :param name:        The name of the label.
+        :param ns:          The namespace to register label to.
+
+        :return: 
         """
-        Returns an address to a given label
-        To show all available labels, iterate over .labels or list(b.kb.labels)
+        namespace = self._namespaces[ns]  # a shorthand
+
+        # Name is already present in the namespace.
+        # Here we forbid assigning different addresses with the same name.
+        if name in namespace and namespace[name] != addr:
+            raise ValueError("Label '%s' is already in present in the namespace "
+                             "'%s' with an address %#x, different from %#x"
+                             % (name, ns, namespace[name], addr))
+
+        # Specified address is already labeled.
+        # Here we forbid assigning different labels to a single address.
+        if addr in namespace.inv and namespace.inv[addr] != name:
+            raise ValueError("Address %#x is already labeled with '%s' "
+                             "in the namespace '%s'" % (addr, name, ns))
+
+        # Create a new label!
+        namespace[name] = addr
+
+    def get_label(self, addr, ns=_global_ns, default=None):
+        """Get a label that is present within the given namespace and is assigned to the specified address.
+        
+        :param addr:    The address for which the label is assigned.
+        :param ns:      The namespace to look into.
+        :param default: The name of the label to assign to the address, if there is no label present.
+                        This accepts the following values: 
+                            True - assign a new label with default name; 
+                            None - don't do anything, i.e. return None if the address is not labeled; 
+                            Any other value - create a new label with the given value.
+        :return: 
         """
-        return self._reverse_labels[name]
+        namespace = self._namespaces[ns]  # a shorthand
 
-    def copy(self):
-        o = Labels(self._kb)
-        o._labels = {k: v for k, v in self._labels.iteritems()}
-        o._reverse_labels = {k: v for k, v in self._reverse_labels.iteritems()}
+        if addr not in namespace.inv and default:
+            if default is True:
+                default = self._generate_default_label(addr)
+            return namespace.setdefault(default, addr)
+
+        else:
+            return namespace.inv.get(addr)
+
+    def del_label(self, name, ns=_global_ns):
+        """Delete a label that is present within the given namespace and is assigned to the specified address.
```

Comment:
Based on the docstring ("and is assigned to the specified address"), did you mean to also require an address in the args here? If label names are unique within a namespace, I don't think it's necessary, though it would be nice to have a similarly simple way to delete or reset-to-default by address.

Reply:
Thanks for notice! No, it is a mistake in the doc-string. Address is not required here, as this function is meant to delete label given it's name only. The doc-string will be updated in the next commits.